core/consensus/babe/src/lib.rs

```diff
@@ -75,28 +157,949 @@ impl Decode for BabeSeal {
+						"hash_previously" => ?pre_hash
+					);
+
+					if let Err(e) = block_import.import_block(import_block, Default::default()) {
+						warn!(target: "babe", "Error with block built on {:?}: {:?}",
+							  parent_hash, e);
+						telemetry!(CONSENSUS_WARN; "babe.err_with_block_built_on";
+							"hash" => ?parent_hash, "err" => ?e
+						);
+					}
+				})
+				.map_err(|e| {
+					warn!("Client import failed: {:?}", e);
+					consensus_common::ErrorKind::ClientImport(format!("{:?}", e)).into()
+				})
+		)
+	}
+}
+
+/// check a header has been signed by the right key. If the slot is too far in
+/// the future, an error will be returned. If successful, returns the pre-header
+/// and the digest item containing the seal.
+///
+/// This digest item will always return `Some` when used with `as_babe_seal`.
+//
+// FIXME #1018 needs misbehavior types
+#[forbid(warnings)]
+fn check_header<B: Block + Sized>(
+	slot_now: u64,
+	mut header: B::Header,
+	hash: B::Hash,
+	authorities: &[Public],
+	threshold: u64,
+) -> Result<CheckedHeader<B::Header, DigestItemFor<B>>, String>
+	where DigestItemFor<B>: CompatibleDigestItem,
+{
+	trace!(target: "babe", "Checking header");
+	let digest_item = match header.digest_mut().pop() {
+		Some(x) => x,
+		None => return Err(format!("Header {:?} is unsealed", hash)),
+	};
+
+	let BabeSeal {
+		slot_num,
+		signature: LocalizedSignature {signer, signature },
+		proof,
+		vrf_output,
+	} = digest_item.as_babe_seal().ok_or_else(|| {
+		error!(target: "babe", "Header {:?} is unsealed", hash);
+		format!("Header {:?} is unsealed", hash)
+	})?;
+
+	if slot_num > slot_now {
+		header.digest_mut().push(digest_item);
+		Ok(CheckedHeader::Deferred(header, slot_num))
+	} else if !authorities.contains(&signer) {
+		error!(target: "babe", "Slot Author not found");
+		Err("Slot Author not found".to_string())
+	} else {
+		let pre_hash = header.hash();
+		let to_sign = (slot_num, pre_hash, proof.to_bytes()).encode();
+
+		if sr25519::Pair::verify(&signature, &to_sign[..], &signer) {
+			let (inout, _batchable_proof) = {
+				let transcript = make_transcript(
+					Default::default(),
+					slot_num,
+					Default::default(),
+					0,
+				);
+				schnorrkel::PublicKey::from_bytes(signer.as_slice()).and_then(|p| {
+					p.vrf_verify(transcript, &vrf_output, &proof)
+				}).map_err(|s| {
+					warn!(target: "babe", "VRF verification failed: {:?}", s);
+					format!("VRF verification failed")
+				})?
+			};
+			if check(&inout, threshold) {
+				Ok(CheckedHeader::Checked(header, digest_item))
+			} else {
+				error!(target: "babe", "VRF verification failed: threshold {} exceeded", threshold);
+				Err(format!("Validator {:?} made seal when it wasn’t its turn", signer))
+			}
+		} else {
+			error!(target: "babe", "Bad signature on {:?}", hash);
+			Err(format!("Bad signature on {:?}", hash))
+		}
+	}
+}
+
+/// Extra verification for Babe blocks.
+pub trait ExtraVerification<B: Block>: Send + Sync {
```

Comment:
I wonder if we could extract this into the `consensus::Common` basic-queue set of tools (re-exporting from Aura for backcompat)

Reply:
I think we can.